From 709e562023b50d56f826f4b8a614340f44cef561 Mon Sep 17 00:00:00 2001
From: Stefan Roesch <shr@devkernel.io>
Date: Wed, 16 Nov 2022 13:09:14 -0800
Subject: [RFC PATCH v4 3/3] io_uring: add api to set napi prefer busy poll

This adds an api to register and unregister the napi prefer busy poll
setting from liburing. To be able to use this functionality, the
corresponding liburing patch is needed.

Signed-off-by: Stefan Roesch <shr@devkernel.io>
---
 include/uapi/linux/io_uring.h | 3 ++-
 io_uring/io_uring.c           | 5 ++++-
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/include/uapi/linux/io_uring.h b/include/uapi/linux/io_uring.h
index 1a713bbafaee..b74021454bf7 100644
--- a/include/uapi/linux/io_uring.h
+++ b/include/uapi/linux/io_uring.h
@@ -619,7 +619,8 @@ struct io_uring_buf_reg {
 /* argument for IORING_(UN)REGISTER_NAPI */
 struct io_uring_napi {
 	__u32	busy_poll_to;
-	__u32	pad;
+	__u8    prefer_busy_poll;
+	__u8	pad[3];
 	__u64	resv;
 };
 
diff --git a/io_uring/io_uring.c b/io_uring/io_uring.c
index cf0e7cc8ad2e..dd903c71fa20 100644
--- a/io_uring/io_uring.c
+++ b/io_uring/io_uring.c
@@ -4127,6 +4127,7 @@ static int io_register_napi(struct io_ring_ctx *ctx, void __user *arg)
 #ifdef CONFIG_NET_RX_BUSY_POLL
 	const struct io_uring_napi curr = {
 		.busy_poll_to = ctx->napi_busy_poll_to,
+		.prefer_busy_poll = ctx->napi_prefer_busy_poll
 	};
 	struct io_uring_napi *napi;
 
@@ -4135,7 +4136,7 @@ static int io_register_napi(struct io_ring_ctx *ctx, void __user *arg)
 		return PTR_ERR(napi);
 
 	WRITE_ONCE(ctx->napi_busy_poll_to, napi->busy_poll_to);
-
+	WRITE_ONCE(ctx->napi_prefer_busy_poll, !!napi->prefer_busy_poll);
 	kfree(napi);
 
 	if (copy_to_user(arg, &curr, sizeof(curr)))
@@ -4152,12 +4153,14 @@ static int io_unregister_napi(struct io_ring_ctx *ctx, void __user *arg)
 #ifdef CONFIG_NET_RX_BUSY_POLL
 	const struct io_uring_napi curr = {
 		.busy_poll_to = ctx->napi_busy_poll_to,
+		.prefer_busy_poll = ctx->napi_prefer_busy_poll
 	};
 
 	if (copy_to_user(arg, &curr, sizeof(curr)))
 		return -EFAULT;
 
 	WRITE_ONCE(ctx->napi_busy_poll_to, 0);
+	WRITE_ONCE(ctx->napi_prefer_busy_poll, false);
 	return 0;
 #else
 	return -EINVAL;
-- 
2.30.2

